src/lib/uri/uri.c

```diff
@@ -349,3 +350,91 @@ uri_set_create(struct uri_set *uri_set, const char *str)
 	uri_set->uri_count = 0;
 	return -1;
 }
+
+static char url_encode_table[256] = {0};
+
+/*
+ * Builds an array with unreserved characters.
+ */
+static void
+build_url_encode_table(const char *str)
+{
+	for (; *str; str++) {
+		int ch = (int)*str;
+		url_encode_table[ch] = 1;
+	}
+}
+
+char *
+str_escape(const char *buf, const char *unreserved, bool encode_plus)
+{
+	build_url_encode_table(unreserved);
+	size_t len = strlen(buf);
+	/* The worst case is when all characters are encoded. */
+	char *encoded = xcalloc(len * 3 + 1, sizeof(char));
```

Comment:
If this code is written for performance, there are several obvious things to improve:

1. Building the same table for each call. It is easy to avoid at least for standard (and most used) character sets.
2. Using `malloc()`/`calloc()` for a temporary buffer (at least it is temporary when we call it from Lua, `ffi.string()` copies its content into a new memory area).
3. Using `strlen()` for a string with known length (it'll also work incorrectly on data with zero bytes inside).
4. Using `calloc()` instead of `*encoded_ptr = '\0';` at the end (and it is also unnecessary for Lua, `ffi.string()` accepts a length).

If we aim two goals: an implementation for using in C and a helper for the Lua implementation, I need to think a bit more, but at least some points are applicable. Generally we prefer to work with a user provided buffer.

My point here is that we should either prefer a simple implementation on this step or play with performance seriously. If a performance is the goal it'll also require considerable amount of my time to review for optimization opportunities (if we pay simplicity, we should gain something really good in speed).

I would also say that pure Lua implementation has several optimization opportunities, which may be validated. At least, precalculating patterns. If we don't aim performance here I can overlook it in my review.

Reply:
Fun fact: when the code was adjusted to work with zero bytes inside source/destination/unreserved strings, the recommendations above were implicitly applied.